contracts/ckb-cheque-script/src/entry.rs

```diff
@@ -23,24 +21,34 @@ pub fn main() -> Result<(), Error> {
         return Err(Error::InvalidArgument);
     }
     
-    // The receiver may be lock args or lock hash
-    let mut receiver = [0u8; 20];
+    let mut receiver_lock_hash = [0u8; 20];
     let mut sender_lock_hash = [0u8; 20];
-    receiver.copy_from_slice(&args[0..20]);
+    receiver_lock_hash.copy_from_slice(&args[0..20]);
     sender_lock_hash.copy_from_slice(&args[20..]);
 
     let cheque_witness_is_none = cheque_cell_witness_is_none()?;
-    let receiver_has_same_input = helper::has_input_by_lock_hash(receiver);
-    let sender_has_same_input = helper::has_input_by_lock_hash(sender_lock_hash);
-    let is_claim = !cheque_witness_is_none || (cheque_witness_is_none && receiver_has_same_input);
-    let is_withdraw = cheque_witness_is_none && sender_has_same_input;
-
-    return if is_claim {
-        claim::validate(receiver, cheque_witness_is_none, sender_lock_hash)
-    } else if is_withdraw {
-        withdraw::validate(sender_lock_hash)
+    if cheque_witness_is_none {
+        // Check if the inputs contain the same input as receiver lock hash or sender lock hash
+        if helper::has_input_by_lock_hash(&receiver_lock_hash) {
+            claim::validate(&sender_lock_hash, &receiver_lock_hash, true)
+        } else if helper::has_input_by_lock_hash(&sender_lock_hash) {
+            withdraw::validate(&sender_lock_hash, true)
+        } else {
+            Err(Error::NoMatchedInputs)
+        }
     } else {
-        Err(Error::NoMatchedInputs)
+        // Validate the signatures of receiver and sender
+        let mut context = unsafe{ CKBDLContext::<[u8; 128 * 1024]>::new()};
+        let lib = LibSecp256k1::load(&mut context);
+        match helper::validate_blake2b_sighash_all(&lib, &receiver_lock_hash) {
+            Ok(_) => claim::validate(&sender_lock_hash, &receiver_lock_hash, false),
```

Comment:
This means that if it is not a claim, it is withdraw, right?

Reply:
Yes

Reply:
OK